src/github.com/getlantern/http-proxy/server.go

```diff
@@ -164,12 +183,15 @@ func (s *Server) doServe(listener net.Listener, chListenOn *chan string) error {
 
 	s.httpServer = http.Server{Handler: proxy,
 		ConnState: func(c net.Conn, state http.ConnState) {
-			if state == http.StateActive {
-				select {
-				case q <- c:
-				default:
-					fmt.Print("Oops! the connection queue is full!\n")
-				}
+			switch state {
+			case http.StateActive:
+				cb.Put(c)
+			case http.StateClosed:
```

Comment:
I may chose a wrong name, but `connBag.Withdraw()` which invoked as the very first statement in handler will also remove it. `Purge` here only takes effect when the handler didn't get called.